rust/src/progress.rs

```diff
@@ -0,0 +1,152 @@
+            ProgressType::NItems(n as u64)
+        } else {
+            ProgressType::Task
+        };
+        let target = ProgressDrawTarget::stderr();
+        let style = ProgressStyle::default_bar();
+        let pb = match ptype {
+            ProgressType::Task => {
+                let pb = ProgressBar::new_spinner();
+                pb.set_style(style.template("{spinner} {prefix} {msg}"));
+                pb.enable_steady_tick(200);
+                pb
+            }
+            ProgressType::NItems(n) => {
+                let pb = ProgressBar::new(n);
+                pb.set_style(
+                    style.template("{spinner} {prefix} {pos}/{len} [{bar:20}] ({eta}) {msg}"),
+                );
+                pb
+            }
+            ProgressType::Percent => {
+                let pb = ProgressBar::new(100);
+                pb.set_style(style.template("{spinner} {prefix} {pos}% [{bar:20}] ({eta}) {msg}"));
+                pb
+            }
+        };
+        let is_hidden = target.is_hidden();
+        if is_hidden {
+            print!("{}...", msg);
+        } else {
+            let msg = match ptype {
+                ProgressType::Task => Cow::Owned(format!("{}...", msg)),
+                _ => Cow::Borrowed(&msg),
+            };
+            pb.set_prefix(&msg);
+        }
+        *lock = Some(ProgressState {
+            bar: pb,
+            is_hidden: is_hidden,
+            ptype: ptype,
+            message: msg,
+        });
+    }
+
+    /// Update progress bar
+    #[no_mangle]
+    pub extern "C" fn ror_progress_set_message(msg: *const libc::c_char) {
+        let msg = string_from_nonnull(msg);
+        let mut lock = PROGRESS.lock().unwrap();
+        let state = lock.as_mut().expect("progress to update");
+        state.bar.set_prefix(&msg);
+        state.message = msg.to_string();
+    }
+
+    /// Update progress bar
+    #[no_mangle]
+    pub extern "C" fn ror_progress_set_sub_message(msg: *const libc::c_char) {
+        let msg = str_from_nullable(msg);
+        let mut lock = PROGRESS.lock().unwrap();
+        let state = lock.as_mut().expect("progress to update");
+        if let Some(ref sub_message) = msg {
+            state.bar.set_message(sub_message)
+        } else {
+            state.bar.set_message("");
+        }
+    }
+
+    /// Update progress bar
+    #[no_mangle]
+    pub extern "C" fn ror_progress_update(n: libc::c_int) {
+        let lock = PROGRESS.lock().unwrap();
+        let state = lock.as_ref().expect("progress to update");
+        assert!(!(state.ptype == ProgressType::Task));
+        state.bar.set_position(n as u64);
+    }
+
+    #[no_mangle]
+    pub extern "C" fn ror_progress_end() {
+        let mut lock = PROGRESS.lock().unwrap();
+        let state = lock.take().expect("progress to end");
+        state.bar.finish_and_clear();
+        if state.is_hidden {
+            println!("done");
+        } else {
+            println!("{}...done", state.message);
```

Comment:
OK, after reading below... this could trivially just take a text string for the "done" part, right? E.g. the pkgcache migration bit prints out the number of pkgs migrated.